File: backend/analysis/fundamental.py
"""
Fundamental analysis scoring engine.

Reads the most recent Fundamental record for a ticker from the DB, scores
it across four dimensions (value, quality, growth, dividend), and returns
a weighted total.
"""

from __future__ import annotations

import logging
from typing import Optional

import numpy as np
from sqlalchemy import select
from sqlalchemy.orm import Session

from db.models import Fundamental

logger = logging.getLogger(__name__)

# ── Weight configuration ─────────────────────────────────────────────────────
WEIGHTS = {
    "value": 0.25,
    "quality": 0.30,
    "growth": 0.30,
    "dividend": 0.15,
}


# ── Threshold-based scoring helpers ──────────────────────────────────────────

def _score_metric(
    value: Optional[float],
    thresholds: list[tuple[float, float]],
    higher_is_better: bool = True,
    default: float = 50.0,
) -> float:
    """Score a single metric against a list of (threshold, score) pairs.

    *thresholds* is an ascending list of ``(threshold_value, score)`` tuples.
    If ``higher_is_better`` is True the highest matching threshold wins;
    otherwise the lowest matching threshold wins.
    """
    if value is None or np.isnan(value):
        return default

    result = default
    for threshold, score in thresholds:
        if higher_is_better:
            if value >= threshold:
                result = score
        else:
            if value <= threshold:
                result = score
    return result


# ── Sub-score functions ──────────────────────────────────────────────────────

def _value_score(f: Fundamental) -> float:
    """Assess valuation cheapness (0-100, higher = cheaper/better)."""
    components: list[float] = []

    # P/E ratio: lower is better
    pe = f.pe_ratio
    if pe is not None and not np.isnan(pe) and pe > 0:
        if pe < 10:
            components.append(90)
        elif pe < 15:
            components.append(75)
        elif pe < 20:
            components.append(60)
        elif pe < 30:
            components.append(40)
        else:
            components.append(20)
    else:
        components.append(50)

    # P/B ratio: lower is better
    pb = f.pb_ratio
    if pb is not None and not np.isnan(pb) and pb > 0:
        if pb < 1.0:
            components.append(90)
        elif pb < 2.0:
            components.append(70)
        elif pb < 4.0:
            components.append(50)
        else:
            components.append(25)
    else:
        components.append(50)

    # P/S ratio
    ps = f.ps_ratio
    if ps is not None and not np.isnan(ps) and ps > 0:
        if ps < 1.0:
            components.append(85)
        elif ps < 3.0:
            components.append(65)
        elif ps < 6.0:
            components.append(45)
        else:
            components.append(20)
    else:
        components.append(50)

    # PEG ratio
    peg = f.peg_ratio
    if peg is not None and not np.isnan(peg) and peg > 0:
        if peg < 1.0:
            components.append(85)
        elif peg < 1.5:
            components.append(65)
        elif peg < 2.5:
            components.append(45)
        else:
            components.append(25)
    else:
        components.append(50)

    # EV/EBITDA
    ev = f.ev_to_ebitda
    if ev is not None and not np.isnan(ev) and ev > 0:
        if ev < 8:
            components.append(85)
        elif ev < 12:
            components.append(65)
        elif ev < 18:
            components.append(45)
        else:
            components.append(25)
    else:
        components.append(50)

    return float(np.mean(components))


def _quality_score(f: Fundamental) -> float:
    """Assess business quality (0-100, higher = better)."""
    components: list[float] = []

    # ROE
    roe = f.roe
    if roe is not None and not np.isnan(roe):
        if roe > 25:
            components.append(90)
        elif roe > 15:
            components.append(75)
        elif roe > 10:
            components.append(55)
        elif roe > 0:
            components.append(35)
        else:
            components.append(15)
    else:
        components.append(50)

    # ROA
    roa = f.roa
    if roa is not None and not np.isnan(roa):
        if roa > 15:
            components.append(90)
        elif roa > 8:
            components.append(70)
        elif roa > 3:
            components.append(50)
        elif roa > 0:
            components.append(30)
        else:
            components.append(15)
    else:
        components.append(50)

    # Debt-to-equity: lower is better
    de = f.debt_to_equity
    if de is not None and not np.isnan(de):
        if de < 0.3:
            components.append(90)
        elif de < 0.7:
            components.append(70)
        elif de < 1.5:
            components.append(50)
        elif de < 3.0:
            components.append(30)
        else:
            components.append(15)
    else:
        components.append(50)

    # Current ratio: higher is safer
    cr = f.current_ratio
    if cr is not None and not np.isnan(cr):
        if cr > 3.0:
            components.append(85)
        elif cr > 2.0:
            components.append(70)
        elif cr > 1.5:
            components.append(55)
        elif cr > 1.0:
            components.append(40)
        else:
            components.append(20)
    else:
        components.append(50)

    # Free cash flow (positive is good; we only check sign here)
    fcf = f.free_cash_flow
    if fcf is not None and not np.isnan(fcf):
        if fcf > 0:
            components.append(70)
        else:
            components.append(25)
    else:
        components.append(50)

    return float(np.mean(components))


def _growth_score(f: Fundamental) -> float:
    """Assess growth profile (0-100, higher = stronger growth)."""
    components: list[float] = []

    # Revenue growth (assumed as decimal, e.g. 0.15 = 15%)
    rg = f.revenue_growth
    if rg is not None and not np.isnan(rg):
        pct = rg * 100 if abs(rg) < 5 else rg  # handle both decimal & percent
        if pct > 30:
            components.append(90)
        elif pct > 15:
            components.append(75)
        elif pct > 5:
            components.append(55)
        elif pct > 0:
            components.append(40)
        else:
            components.append(20)
    else:
        components.append(50)

    # Earnings growth
    eg = f.earnings_growth
    if eg is not None and not np.isnan(eg):
        pct = eg * 100 if abs(eg) < 5 else eg
        if pct > 30:
            components.append(90)
        elif pct > 15:
            components.append(75)
        elif pct > 5:
            components.append(55)
        elif pct > 0:
            components.append(40)
        else:
            components.append(20)
    else:
        components.append(50)

    return float(np.mean(components))


def _dividend_score(f: Fundamental) -> float:
    """Assess dividend attractiveness (0-100)."""
    dy = f.dividend_yield
    if dy is None or np.isnan(dy):
        return 50.0  # no data => neutral

    pct = dy * 100 if dy < 1 else dy  # handle decimal vs percent
    if pct > 5:
        return 85.0
    elif pct > 3:
        return 70.0
    elif pct > 1.5:
        return 55.0
    elif pct > 0:
        return 40.0
    else:
        return 30.0  # zero dividend


# ── Public API ───────────────────────────────────────────────────────────────

def compute_fundamental_score(ticker: str, session: Session) -> dict:
    """Compute a weighted fundamental score for *ticker*.

    Returns a dict with sub-scores and total, or an error dict when no
    fundamental data is available.
    """
    fund: Optional[Fundamental] = (
        session.execute(
            select(Fundamental)
            .where(Fundamental.ticker == ticker)
            .order_by(Fundamental.date_fetched.desc())
        )
        .scalars()
        .first()
    )

    if fund is None:
        logger.warning("No fundamental data for %s", ticker)
        return {"ticker": ticker, "error": "no_fundamental_data"}

    value = _value_score(fund)
    quality = _quality_score(fund)
    growth = _growth_score(fund)
    dividend = _dividend_score(fund)

    total = (
        value * WEIGHTS["value"]
        + quality * WEIGHTS["quality"]
        + growth * WEIGHTS["growth"]
        + dividend * WEIGHTS["dividend"]
    )

    return {
        "ticker": ticker,
        "date_fetched": str(fund.date_fetched),
        "value_score": round(value, 2),
        "quality_score": round(quality, 2),
        "growth_score": round(growth, 2),
        "dividend_score": round(dividend, 2),
        "fundamental_score": round(total, 2),
        "sector": fund.sector,
        "industry": fund.industry,
    }

File: backend/src/analysis/fundamental.ts
/**
 * Fundamental analysis scoring engine.
 * Scores four dimensions (value, quality, growth, dividend) and returns a weighted total.
 */

import { PrismaClient, Fundamental } from '@prisma/client';
import { getPrisma } from '../db/client';
import { mean as arrMean } from '../utils/math';

const WEIGHTS = {
  value: 0.25,
  quality: 0.30,
  growth: 0.30,
  dividend: 0.15,
};

function isValid(val: number | null | undefined): val is number {
  return val !== null && val !== undefined && !isNaN(val);
}

function valueScore(f: Fundamental): number {
  const components: number[] = [];

  // PE ratio: lower is better
  const pe = f.peRatio;
  if (isValid(pe) && pe > 0) {
    if (pe < 10) components.push(90);
    else if (pe < 15) components.push(75);
    else if (pe < 20) components.push(60);
    else if (pe < 30) components.push(40);
    else components.push(20);
  } else components.push(50);

  // PB ratio: lower is better
  const pb = f.pbRatio;
  if (isValid(pb) && pb > 0) {
    if (pb < 1.0) components.push(90);
    else if (pb < 2.0) components.push(70);
    else if (pb < 4.0) components.push(50);
    else components.push(25);
  } else components.push(50);

  // PS ratio
  const ps = f.psRatio;
  if (isValid(ps) && ps > 0) {
    if (ps < 1.0) components.push(85);
    else if (ps < 3.0) components.push(65);
    else if (ps < 6.0) components.push(45);
    else components.push(20);
  } else components.push(50);

  // PEG ratio
  const peg = f.pegRatio;
  if (isValid(peg) && peg > 0) {
    if (peg < 1.0) components.push(85);
    else if (peg < 1.5) components.push(65);
    else if (peg < 2.5) components.push(45);
    else components.push(25);
  } else components.push(50);

  // EV/EBITDA
  const ev = f.evToEbitda;
  if (isValid(ev) && ev > 0) {
    if (ev < 8) components.push(85);
    else if (ev < 12) components.push(65);
    else if (ev < 18) components.push(45);
    else components.push(25);
  } else components.push(50);

  return arrMean(components);
}

function qualityScore(f: Fundamental): number {
  const components: number[] = [];

  // ROE
  const roe = f.roe;
  if (isValid(roe)) {
    if (roe > 25) components.push(90);
    else if (roe > 15) components.push(75);
    else if (roe > 10) components.push(55);
    else if (roe > 0) components.push(35);
    else components.push(15);
  } else components.push(50);

  // ROA
  const roa = f.roa;
  if (isValid(roa)) {
    if (roa > 15) components.push(90);
    else if (roa > 8) components.push(70);
    else if (roa > 3) components.push(50);
    else if (roa > 0) components.push(30);
    else components.push(15);
  } else components.push(50);

  // Debt-to-equity: lower is better
  const de = f.debtToEquity;
  if (isValid(de)) {
    if (de < 0.3) components.push(90);
    else if (de < 0.7) components.push(70);
    else if (de < 1.5) components.push(50);
    else if (de < 3.0) components.push(30);
    else components.push(15);
  } else components.push(50);

  // Current ratio
  const cr = f.currentRatio;
  if (isValid(cr)) {
    if (cr > 3.0) components.push(85);
    else if (cr > 2.0) components.push(70);
    else if (cr > 1.5) components.push(55);
    else if (cr > 1.0) components.push(40);
    else components.push(20);
  } else components.push(50);

  // Free cash flow
  const fcf = f.freeCashFlow;
  if (isValid(fcf)) {
    components.push(fcf > 0 ? 70 : 25);
  } else components.push(50);

  return arrMean(components);
}

function growthScore(f: Fundamental): number {
  const components: number[] = [];

  for (const rawVal of [f.revenueGrowth, f.earningsGrowth]) {
    if (isValid(rawVal)) {
      const pct = Math.abs(rawVal) < 5 ? rawVal * 100 : rawVal;
      if (pct > 30) components.push(90);
      else if (pct > 15) components.push(75);
      else if (pct > 5) components.push(55);
      else if (pct > 0) components.push(40);
      else components.push(20);
    } else components.push(50);
  }

  return arrMean(components);
}

function dividendScore(f: Fundamental): number {
  const dy = f.dividendYield;
  if (!isValid(dy)) return 50;

  const pct = dy < 1 ? dy * 100 : dy;
  if (pct > 5) return 85;
  if (pct > 3) return 70;
  if (pct > 1.5) return 55;
  if (pct > 0) return 40;
  return 30;
}

export async function computeFundamentalScore(
  ticker: string,
  prisma?: PrismaClient,
): Promise<Record<string, any>> {
  const db = prisma || getPrisma();

  const fund = await db.fundamental.findFirst({
    where: { ticker },
    orderBy: { dateFetched: 'desc' },
  });

  if (!fund) {
    console.warn(`No fundamental data for ${ticker}`);
    return { ticker, error: 'no_fundamental_data' };
  }

  const value = valueScore(fund);
  const quality = qualityScore(fund);
  const growth = growthScore(fund);
  const dividend = dividendScore(fund);

  const total =
    value * WEIGHTS.value +
    quality * WEIGHTS.quality +
    growth * WEIGHTS.growth +
    dividend * WEIGHTS.dividend;

  return {
    ticker,
    date_fetched: fund.dateFetched.toISOString().split('T')[0],
    value_score: Math.round(value * 100) / 100,
    quality_score: Math.round(quality * 100) / 100,
    growth_score: Math.round(growth * 100) / 100,
    dividend_score: Math.round(dividend * 100) / 100,
    fundamental_score: Math.round(total * 100) / 100,
    sector: fund.sector,
    industry: fund.industry,
  };
}
